File: test_regress/t/t_verilated_all.pl
#!/usr/bin/env perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2003-2009 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

scenarios(vltmt => 1);

my $root = "..";

compile(
    # Can't use --coverage and --savable together, so cheat and compile inline
    verilator_flags2 => ["--cc",
                         "--coverage-toggle --coverage-line --coverage-user",
                         "--trace --vpi ",
                         "--trace-threads 1",
                         "--prof-exec", "--prof-pgo",
                         "$root/include/verilated_save.cpp"],
    threads => 2
    );

execute(
    all_run_flags => [" +verilator+prof+exec+file+/dev/null",
                      " +verilator+prof+vlt+file+/dev/null",
                      ],
    check_finished => 1,
    );

my %hit;
foreach my $file (glob("$root/include/*.cpp $root/include/*.h")) {
    $file =~ s!.*/!!;

    # This file isn't actually used by the runtime (though
    # it might be in the future? hence it's under include/)
    # It is used to build verilator.
    if ($file =~ /verilated_unordered_set_map\.h/) { next; }

    print "NEED: $file\n" if $Self->{verbose};
    $hit{$file} = 0;
}
foreach my $dfile (glob("$Self->{obj_dir}/*.d")) {
    my $wholefile = file_contents($dfile);
    foreach my $file (split /\s+/, $wholefile) {
        $file =~ s!.*/!!;
        print "USED: $file\n" if $Self->{verbose};
        $hit{$file} = 1;
    }
}

foreach my $file (sort keys %hit) {
    if (!$hit{$file}
        && $file !~ /_sc/
        && $file !~ /_fst/
        && $file !~ /_heavy/
        && ($file !~ /_thread/)) {
        error("Include file not covered by t_verilated_all test: ", $file);
    }
}

ok(1);
1;

File: test_regress/t/t_verilated_all.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of either the GNU Lesser General Public License Version 3
# or the Perl Artistic License Version 2.0.
# SPDX-FileCopyrightText: 2024 Wilson Snyder
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('vltmt')

if not os.path.exists(test.root + "/.git"):
    test.skip("Not in a git repository")

if not re.search(r'g\+\+|GCC|clang', test.cxx_version):
    test.skip("Compiler does not support -fno-exceptions")

test.compile(
    # Can't use --coverage and --savable together, or multiple trace formats, so cheat and compile inline
    verilator_flags2=[
        "--cc",
        "--coverage-toggle --coverage-line --coverage-user",
        "--trace-fst",  # Also adds -lz4 and the like
        "--vpi",
        ("--timing" if test.have_coroutines else "--no-timing -Wno-STMTDLY"),
        "--prof-exec",
        "--prof-pgo",
        # -fno-exceptions checks the runtime/generated code does not rely on exceptions
        "-CFLAGS -fno-exceptions",
        test.root + "/include/verilated_save.cpp",
        test.root + "/include/verilated_vcd_c.cpp",
        test.root + "/include/verilated_saif_c.cpp"
    ],
    threads=2)

test.execute(
    all_run_flags=[" +verilator+prof+exec+file+/dev/null", " +verilator+prof+vlt+file+/dev/null"])

hit = {}
for filename in (test.glob_some(test.root + "/include/*.cpp") +
                 test.glob_some(test.root + "/include/*.h")):
    filename = os.path.basename(filename)
    if test.verbose:
        print("NEED: " + filename)
    hit[filename] = False

for dfile in test.glob_some(test.obj_dir + "/*.d"):
    wholefile = test.file_contents(dfile)
    for filename in wholefile.split():
        filename = os.path.basename(filename)
        if test.verbose:
            print("USED: " + filename)
        hit[filename] = True

for filename in sorted(hit.keys()):
    if hit[filename]:
        continue
    if re.search(r'_sc', filename):  # SystemC files are not built by this test
        continue
    if re.search(r'_timing', filename) and not test.have_coroutines:
        continue
    test.error("Include file not covered by t_verilated_all test: ", filename)

test.passes()
